File: YuXiangROS/Chap2/2.4_2.5_Workspace/src/demo_cpp_pkg/src/person_node.cpp
/*
1. 一包多节点: 在一个包中可以包含多个节点, 在CMakeLists.txt中添加:
	add_executable(person_node src/person_node.cpp)
	target_link_libraries(person_node rclcpp)
	install(TARGETS person_node DESTINATION lib/${PROJECT_NAME})

2. 节点类化: 将节点封装为类, 便于管理
*/

#include <rclcpp/rclcpp.hpp>
#include <string>

class PersonNode : public rclcpp::Node {
  private:
    std::string name_;
    int age_;

  public:
    PersonNode(const std::string& node_name, const std::string& name, const int& age)
        : Node(node_name) {
        this->name_ = name;
        this->age_ = age;
    }

    void say_hello(const std::string& hobby) {
        RCLCPP_INFO(
            this->get_logger(),
            "Hello, my name is %s, and I am %d years old. I like %s.",
            this->name_.c_str(),
            this->age_,
            hobby.c_str()
        );
    } // .c_str() 将string转换为const char*
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto person_node = std::make_shared<PersonNode>("person_node", "ChangLi", 20);
    person_node->say_hello("reading");
    rclcpp::spin(person_node);
    rclcpp::shutdown();
    return 0;
}

File: YuXiangROS/Chap2/2.4_2.5_Workspace/src/demo_python_pkg/demo_python_pkg/person_node.py
"""
1. 一包多节点: 在一个包中可以包含多个节点, 在 setup.py 中的 entry_points 中添加多个节点即可.
2. 节点类化: 将节点封装为类, 便于管理
"""

import rclpy
from rclpy.node import Node


class PersonNode(Node):
    def __init__(self, node_name: str, name: str, age: int) -> None:
        super().__init__(node_name)

        self.name = name
        self.age = age

    def say_hello(self, hobby: str):
        self.get_logger().info(f"Hello, my name is {self.name}, and I am {self.age} years old. I like {hobby}.")


def main(args=None):
    rclpy.init(args=args)
    person_node = PersonNode("person_node", "Changli", 20)
    person_node.say_hello("reading")
    try:
        rclpy.spin(person_node)
    except KeyboardInterrupt:
        pass
    finally:
        person_node.destroy_node()
        rclpy.shutdown()
